xla/service/gpu/analytical_latency_estimator_test.cc

```diff
@@ -0,0 +1,167 @@
+      return ShapeUtil::ByteSizeOf(shape, kPointerSize);
+    };
+  }
+};
+
+TEST_F(AnalyticalLatencyHidingSchedulerTest, TestAnalyticalLatencyEstimator) {
+  // On pre-Ampere GPUs the test would use a different amount of shared memory.
+  if (!GetCudaComputeCapability().IsAtLeast(
+          se::CudaComputeCapability::PASCAL_)) {
+    GTEST_SKIP() << "This test is for Pascal+ GPUs.";
+  }
+  const GpuDeviceInfo dev_info =
+      GetGpuDeviceInfo(backend().default_stream_executor());
+
+  absl::string_view hlo_string = R"(
+HloModule module, is_scheduled=true
+
+region_20.995 {
+  Arg_1.997 = f32[] parameter(1)
+  Arg_0.996 = f32[] parameter(0)
+  ROOT add.589 = f32[] add(Arg_0.996, Arg_1.997)
+}
+
+ENTRY entry {
+  p0 = f32[16,64,256]{2,1,0} parameter(0)
+  p1 = f32[16,64,256]{2,1,0} parameter(1)
+  p2 = f32[1024,2048,2048]{2,1,0} parameter(2)
+  p3 = f32[2048,2048,2048]{2,1,0} parameter(3)
+  all-reduce-start.1 = f32[1024,2048,2048]{2,1,0} all-reduce-start(p2), channel_id=8, replica_groups={{0}}, to_apply=region_20.995, backend_config="{\"is_sync\":false}"
+  all-reduce-start.2 = f32[2048,2048,2048]{2,1,0} all-reduce-start(p3), channel_id=10, replica_groups={{0}}, to_apply=region_20.995, backend_config="{\"is_sync\":false}"
+
+  all-reduce-done.1 = f32[1024,2048,2048]{2,1,0} all-reduce-done(all-reduce-start.1)
+  all-reduce-done.2 = f32[2048,2048,2048]{2,1,0} all-reduce-done(all-reduce-start.2)
+  conv0 = f32[16,256,256]{2,1,0} convolution(p0, p1),
+    window={size=16 stride=15 lhs_dilate=16}, dim_labels=0fb_0io->0fb
+
+  ROOT tuple.2 = (f32[16,256,256]{2,1,0}, f32[1024,2048,2048]{2,1,0}, f32[2048,2048,2048]{2,1,0}) tuple(conv0, all-reduce-done.1, all-reduce-done.2)
+}
+)";
+
+  TF_ASSERT_OK_AND_ASSIGN(auto hlo_module, ParseHloText(hlo_string));
+  hlo_module->config().set_num_partitions(2);
+
+  HloSchedule& module_schedule = hlo_module->schedule();
+  EXPECT_TRUE(hlo_module->has_entry_computation());
+
+  auto scheduler_config = GetDefaultSchedulerConfig();
+  auto latency_estimator = std::make_unique<AnalyticalLatencyEstimator>(
+      scheduler_config, std::make_unique<ApproximateLatencyEstimator>(),
+      dev_info, ShapeSizeBytesFunction(), hlo_module->entry_computation());
+  EXPECT_TRUE(RunScheduler(hlo_module.get(), scheduler_config,
+                           std::move(latency_estimator))
+                  .ok());
+  EXPECT_TRUE(hlo_module->has_entry_computation());
+
+  std::vector<HloInstruction*> new_instruction_schedule =
+      module_schedule.sequence(hlo_module->entry_computation()).instructions();
+  int64_t ar2_index = GetInstructionIndexInSchedule(new_instruction_schedule,
+                                                    "all-reduce-start.2");
+  int64_t ar1_done_index = GetInstructionIndexInSchedule(
+      new_instruction_schedule, "all-reduce-done.1");
+  int64_t conv0_index =
+      GetInstructionIndexInSchedule(new_instruction_schedule, "conv0");
+
+  EXPECT_LT(ar1_done_index, ar2_index);
```

Comment:
Can you add a comment here or at the start of the test to document what you expect. Here it seems ar2 is 2x the size of ar1, so we expected LHS to overlap the single conv with ar2 as opposed to ar1?

Reply:
added comment